File: 46-permutations/permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        List<Integer> temp = new ArrayList<Integer>();
        result.add(temp);
        for(int i: nums){
            List<List<Integer>> permutations = new ArrayList<List<Integer>>();
            for(List<Integer> t: result){
                for(int j = 0; j < t.size() + 1; j++){
                    List<Integer> tt = new ArrayList<Integer>(t);
                    tt.add(j, i);
                    permutations.add(tt);
                }
            }
            result = permutations;
        }
        return result;
        // if (nums.length == 0){
        //     List<List<Integer>> temp = new ArrayList<List<Integer>>();
        //     List<Integer> temp2 = new ArrayList<Integer>();
        //     temp.add(temp2);
        //     return temp;
        // } 
        // List<List<Integer>> per = permute(Arrays.copyOfRange(nums, 1, nums.length));
        // List<List<Integer>> result = new ArrayList<List<Integer>>();
        // for(int ii: nums) System.out.println(ii);
        // System.out.println(per);
        // for(List<Integer> t: per){
            
        //     for(int j = 0; j < t.size() + 1; j++){
        //         List<Integer> tt = new ArrayList<Integer>(t);
        //         tt.add(j, nums[0]);
        //         result.add(tt);
        //     }
        // }
        // return result;
    }
}

File: 46-permutations/permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result = {{}};
        for(int i: nums){
            vector<vector<int>> permutations;
            for(vector<int> temp: result){
                for(int j=0; j<temp.size() + 1; j++){
                    vector<int> t = temp;
                    t.insert(t.begin() + j, i);
                    permutations.push_back(t);
                }
            }
            result = permutations;
        }
        return result;

        // vector<vector<int>> tt = {{}};
        // if (nums.size() == 0) return tt;
        // vector<int> temp = {nums.begin() + 1, nums.end()};
        // vector<vector<int>> permutation = permute(temp);
        // vector<vector<int>> result;
        // for(vector<int> i: permutation){
        //     for(int j = 0; j < i.size() + 1; j ++){
        //         vector<int> t = i;
        //         t.insert(t.begin() + j, nums[0]);
        //         result.push_back(t);
        //     }
        // }
        // return result;
    }
};